Upgrade ocaml interop

tizoc opened the pull request:
NOTE: this was made on top of #383, which should go first (initially it was the other way around).

Changes:

- Upgraded to newer ocaml-interop 
- Switched to an exception-less tezos-ffi. Tezos errors are represented by the new `TezosError` and `TezosErrorTrace` types.
- Mapping of errors doesn't require any more parsing, instead the name of the exception/error is matched.
- Unwrapped all FFI return values that had the form `Result<Result<T, CustomError>, OcamlError>`, which now look like `Result<T, CustomError>`.
- Initialization and shutdown of the OCaml runtime is handled by the OCaml executor thread
- Removed GC threshold setting
- `OcamlError` has been renamed `OCamlBlockPanic`, and with exceptions not being raised now this should only happen when the block passed to the executor panics.

Matching tezos fork MR: https://gitlab.com/simplestaking/tezos/-/merge_requests/15

## What changed on ocaml-interop

- Switched to a caller-rooting calling convention, which means that now values have to be rooted before passing them to OCaml functions (which now receive `OCamlRef<T>` instead of `OCaml<T>` arguments. That is why you will not see the `ocaml_call!` macro anymore.
- `ocaml_frame!` doesn't greate a `gc` handle anymore, but instead it takes as its first argument a `&mut OCamlRuntime` that has to be passed by the caller. In the case of tezedge, the executor thread creates that handle (upon initialization of the OCaml runtime), and the closure that is passed to the `runtime::execute` function now must accept one `&mut OCamlRuntime` argument that gets passed by `runtime::execute`.
- Now any exception raised by OCaml code will result in a `panic!`. The tezedge-ffi code has been modified to not raise exceptions but instead return `Result<T, ErrorTrace>` values to signal errors. Also `runtime::execute` now returns `Result<T, OCamlBlockPanic>`, with `OCamlBlockPanic` signaling that a panic ocurred inside the executed closure (either raised by Rust code, or a rogue OCaml exception).

tizoc commented on tezos/interop/src/ffi.rs:
```diff
@@ -65,38 +65,39 @@ mod tezos_ffi {
 /// https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#sec467
 pub fn shutdown() {
-    ocaml_interop::OCamlRuntime::shutdown_persistent()
+    // TODO
```
Note, this is done now when the executor thread finishes, but I have to revise all these parts and think more about the management the new runtime handle.

tizoc commented:
Just so leave things ready for when the new ocaml-interop release is out. Maybe I will also include the use of polymorphic variants here to simplify the OCaml side (by removing the need to map polymorphic variants into regular ones).

tizoc commented:
@bkontur some of the drone checks failed (tezos-python-tests, sandbox-launcher-tests), any idea of what the problem could be? http://ci.tezedge.com/simplestaking/tezedge/2253

And seems to be hanged with one of the ocaml-nodes (number 4) containing this message:

```
1 | + ip -o -4 addr show \| awk -F '[ /]+' '/global/ {print $4 ":9734"}' > /artifacts/ocaml_ip4.txt
2 | + rm -f /home/tezos/data/lock
3 | + cp /artifacts/identities/identity_7.json /home/tezos/data/identity.json
4 | + tezos-node config reset --data-dir /home/tezos/data --network ${NETWORK} --no-bootstrap-peers
5 | tezos-node: Error:
6 | Specified --network has chain name
7 | TEZOS_ALPHANET_CARTHAGE_2019-11-28T13:02:13Z, but current configuration
8 | implies expected chain name TEZOS_MAINNET. Use:
9 | tezos-node config init --network <NETWORK> to configure your node.
```

tizoc commented:
This is slower than develop but I am not sure why. The extra time is being spent inside  `apply_block`, the `NextProto.init` call in particular. I haven't figured out why yet.

tizoc commented:
Some timings from Edo's `init`:

```
## this branch + matching tezos fork branch
  
Jan 25 20:13:21.079 - 008-PtEdoTez: Before Main.prepare_first_block
Jan 25 20:15:10.856 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 20:15:12.786 - 008-PtEdoTez: preparing from Genesis
Jan 25 20:15:12.793 - 008-PtEdoTez: After Main.prepare_first_block

## develop

Jan 25 20:22:44.645 - 008-PtEdoTez: Before Main.prepare_first_block
Jan 25 20:22:49.276 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 20:22:51.046 - 008-PtEdoTez: preparing from Genesis
Jan 25 20:22:51.060 - 008-PtEdoTez: After Main.prepare_first_block
```

- First message from [here](https://gitlab.com/simplestaking/tezos/-/blob/tezedge-ffi/src/proto_008_PtEdoTez/lib_protocol/main.ml#L396)
- Second message from [here](https://gitlab.com/simplestaking/tezos/-/blob/tezedge-ffi/src/proto_008_PtEdoTez/lib_protocol/init_storage.ml#L29)

There is nothing happening in between those to other than Lwt's scheduler, but in one case it takes about 100 seconds, while on the other just 5 seconds. It is very confusing, but I will keep investigating tomorrow.

tizoc commented:
Fixed timings with more granularity.

This branch:

```
Jan 25 22:18:57.141 - 008-PtEdoTez: Before prepare_first_block
Jan 25 22:18:57.145 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 22:18:58.026 - 008-PtEdoTez: preparing from Genesis
Jan 25 22:19:30.809 - 008-PtEdoTez: after Commitment_storage.init <==== SLOW (~30s)
Jan 25 22:19:30.811 - 008-PtEdoTez: after Roll_storage.init
Jan 25 22:19:30.823 - 008-PtEdoTez: after Seed_storage.init
Jan 25 22:19:30.830 - 008-PtEdoTez: after Contract_storage.init
Jan 25 22:20:31.335 - 008-PtEdoTez: after Bootstrap_storage.init <==== SLOW (~60s)
Jan 25 22:20:31.478 - 008-PtEdoTez: after Roll_storage.init_first_cycles
Jan 25 22:20:31.514 - 008-PtEdoTez: after Vote_storage.init
Jan 25 22:20:31.517 - 008-PtEdoTez: after Storage.Block_priority.init
Jan 25 22:20:43.835 - 008-PtEdoTez: after Vote_storage.update_listings <=== SLOW (~10s)
Jan 25 22:20:43.835 - 008-PtEdoTez: After prepare_first_block
```

develop

```
Jan 25 22:27:34.860 - 008-PtEdoTez: Before prepare_first_block
Jan 25 22:27:34.861 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 22:27:35.156 - 008-PtEdoTez: preparing from Genesis
Jan 25 22:27:35.884 - 008-PtEdoTez: after Commitment_storage.init
Jan 25 22:27:35.885 - 008-PtEdoTez: after Roll_storage.init
Jan 25 22:27:35.886 - 008-PtEdoTez: after Seed_storage.init
Jan 25 22:27:35.886 - 008-PtEdoTez: after Contract_storage.init
Jan 25 22:27:36.224 - 008-PtEdoTez: after Bootstrap_storage.init
Jan 25 22:27:36.224 - 008-PtEdoTez: after Roll_storage.init_first_cycles
Jan 25 22:27:36.225 - 008-PtEdoTez: after Vote_storage.init
Jan 25 22:27:36.225 - 008-PtEdoTez: after Storage.Block_priority.init
Jan 25 22:27:36.301 - 008-PtEdoTez: after Vote_storage.update_listings
Jan 25 22:27:36.301 - 008-PtEdoTez: After prepare_first_block
```

tizoc commented:
Numbers from inside `Commitment_storage.init`. This function iterates over 30k commitments and initializes each one. To do so it uses an `init` function that performs a `mem` that will error out if the value is already present, followed by a `set`:

```ocaml
let init ctxt k v =
  Context.mem ctxt.context k
  >>= function
  | true ->
      Lwt.return @@ storage_error (Existing_key k)
  | false ->
      Context.set ctxt.context k v >|= fun context -> ok {ctxt with context}
```

What is important is to notice how in the slow version things get slower the more elements have been `set` already.

This isn't calling our storage btw, I added prints and for this particular test I don't get anything during this call (for others I do), so I don't know yet what the issue is.

this branch (notice how things are reasonably fast at first bug get slower as the amount of commitments processed increases):

```
Jan 25 23:05:04.345 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 23:05:04.678 - 008-PtEdoTez: preparing from Genesis
Jan 25 23:05:04.678 - 008-PtEdoTez: commitments = 0
Jan 25 23:05:04.695 - 008-PtEdoTez: commitments = 1000
Jan 25 23:05:04.727 - 008-PtEdoTez: commitments = 2000
Jan 25 23:05:04.799 - 008-PtEdoTez: commitments = 3000
Jan 25 23:05:04.880 - 008-PtEdoTez: commitments = 4000
Jan 25 23:05:04.988 - 008-PtEdoTez: commitments = 5000
Jan 25 23:05:05.127 - 008-PtEdoTez: commitments = 6000
Jan 25 23:05:05.308 - 008-PtEdoTez: commitments = 7000
Jan 25 23:05:05.510 - 008-PtEdoTez: commitments = 8000
Jan 25 23:05:05.757 - 008-PtEdoTez: commitments = 9000
Jan 25 23:05:06.044 - 008-PtEdoTez: commitments = 10000
Jan 25 23:05:06.371 - 008-PtEdoTez: commitments = 11000
Jan 25 23:05:06.715 - 008-PtEdoTez: commitments = 12000
Jan 25 23:05:07.164 - 008-PtEdoTez: commitments = 13000
Jan 25 23:05:07.641 - 008-PtEdoTez: commitments = 14000
Jan 25 23:05:08.231 - 008-PtEdoTez: commitments = 15000
Jan 25 23:05:08.882 - 008-PtEdoTez: commitments = 16000
Jan 25 23:05:09.775 - 008-PtEdoTez: commitments = 17000
Jan 25 23:05:10.741 - 008-PtEdoTez: commitments = 18000
Jan 25 23:05:11.811 - 008-PtEdoTez: commitments = 19000
Jan 25 23:05:12.814 - 008-PtEdoTez: commitments = 20000
Jan 25 23:05:14.090 - 008-PtEdoTez: commitments = 21000
Jan 25 23:05:15.306 - 008-PtEdoTez: commitments = 22000
Jan 25 23:05:16.543 - 008-PtEdoTez: commitments = 23000
Jan 25 23:05:17.951 - 008-PtEdoTez: commitments = 24000
Jan 25 23:05:19.500 - 008-PtEdoTez: commitments = 25000
Jan 25 23:05:21.634 - 008-PtEdoTez: commitments = 26000
Jan 25 23:05:23.615 - 008-PtEdoTez: commitments = 27000
Jan 25 23:05:25.549 - 008-PtEdoTez: commitments = 28000
Jan 25 23:05:27.837 - 008-PtEdoTez: commitments = 29000
Jan 25 23:05:30.287 - 008-PtEdoTez: commitments = 30000
Jan 25 23:05:30.288 - 008-PtEdoTez: after Commitment_storage.init commitments-count=30001
```

develop (fast all the time):

```
Jan 25 23:07:35.292 - 008-PtEdoTez: Before Raw_context.prepare_first_block
Jan 25 23:07:35.586 - 008-PtEdoTez: preparing from Genesis
Jan 25 23:07:35.586 - 008-PtEdoTez: commitments = 0
Jan 25 23:07:35.611 - 008-PtEdoTez: commitments = 1000
Jan 25 23:07:35.632 - 008-PtEdoTez: commitments = 2000
Jan 25 23:07:35.654 - 008-PtEdoTez: commitments = 3000
Jan 25 23:07:35.677 - 008-PtEdoTez: commitments = 4000
Jan 25 23:07:35.699 - 008-PtEdoTez: commitments = 5000
Jan 25 23:07:35.722 - 008-PtEdoTez: commitments = 6000
Jan 25 23:07:35.743 - 008-PtEdoTez: commitments = 7000
Jan 25 23:07:35.771 - 008-PtEdoTez: commitments = 8000
Jan 25 23:07:35.799 - 008-PtEdoTez: commitments = 9000
Jan 25 23:07:35.826 - 008-PtEdoTez: commitments = 10000
Jan 25 23:07:35.851 - 008-PtEdoTez: commitments = 11000
Jan 25 23:07:35.874 - 008-PtEdoTez: commitments = 12000
Jan 25 23:07:35.898 - 008-PtEdoTez: commitments = 13000
Jan 25 23:07:35.919 - 008-PtEdoTez: commitments = 14000
Jan 25 23:07:35.941 - 008-PtEdoTez: commitments = 15000
Jan 25 23:07:35.963 - 008-PtEdoTez: commitments = 16000
Jan 25 23:07:35.993 - 008-PtEdoTez: commitments = 17000
Jan 25 23:07:36.016 - 008-PtEdoTez: commitments = 18000
Jan 25 23:07:36.043 - 008-PtEdoTez: commitments = 19000
Jan 25 23:07:36.071 - 008-PtEdoTez: commitments = 20000
Jan 25 23:07:36.101 - 008-PtEdoTez: commitments = 21000
Jan 25 23:07:36.129 - 008-PtEdoTez: commitments = 22000
Jan 25 23:07:36.156 - 008-PtEdoTez: commitments = 23000
Jan 25 23:07:36.182 - 008-PtEdoTez: commitments = 24000
Jan 25 23:07:36.205 - 008-PtEdoTez: commitments = 25000
Jan 25 23:07:36.229 - 008-PtEdoTez: commitments = 26000
Jan 25 23:07:36.255 - 008-PtEdoTez: commitments = 27000
Jan 25 23:07:36.279 - 008-PtEdoTez: commitments = 28000
Jan 25 23:07:36.301 - 008-PtEdoTez: commitments = 29000
Jan 25 23:07:36.324 - 008-PtEdoTez: commitments = 30000
Jan 25 23:07:36.325 - 008-PtEdoTez: after Commitment_storage.init commitments-count=30001
```

tizoc commented:
Cancelled the last drone build (las commit is just rustfmt), but it had already passed here: http://ci.tezedge.com/simplestaking/tezedge/2359

tizoc commented:
Oh, another detail, this branch was rebased on top of #383

tizoc commented:
And another detail..... you sill see that the roundtrip benchmarks are much much slower. The reason is that on the OCaml side, because the `Ffi_gc` module was removed, I added a call to `Gc.compact ()` after each result is returned, this performs the most expensive GC operation, thats from where all the overhead is coming from. I just changed it to `Gc.minor ()`.

bkontur commented on tezos/api/src/ffi.rs:
```diff
@@ -356,63 +354,61 @@ pub struct ForkingTestchainData {
+    #[fail(display = "Failed to call - message: {:?}!", trace_message)]
     FailedToCall {
-        parsed_error_message: Option<String>,
+        error_id: String,
```
@tizoc 
what about error_id, could be added to display?

tizoc replied:
It is the same as the id that shows up in the first error in `trace_message` (which is a JSON string), but yes, it is better to have it there anyway, will add.

bkontur commented on tezos/api/src/ffi.rs:
```diff
@@ -465,25 +459,17 @@ impl From<CallError> for ApplyBlockError {
+                error_id,
+                trace_message,
+            } => match error_id.as_str() {
+                "ffi.unknown_predecessor_context" => ApplyBlockError::UnknownPredecessorContext {
```
@tizoc 
I am thinking about maybe some module just with this constants,
placed maybe here: tezos/api/src/ocaml_conv/mod.rs as part of a conversion

mod ffi_errors {
 const XYZ: &'static str = "ffi.unknown_predecessor_context"
}

that will match ffi_errors.ml:
https://gitlab.com/simplestaking/tezos/-/merge_requests/15/diffs#5494639cbe5ba8fc72bde25a5b422986f92c6364_12_48

tizoc replied:
Done.

bkontur commented on tezos/api/src/ffi.rs:
```diff
@@ -611,6 +586,7 @@ impl From<CallError> for ValidateOperationError {
     }
 }
 
+// XXX: UNUSED
```
@tizoc 
if unused , please remove

bkontur commented on tezos/api/src/ffi.rs:
```diff
@@ -619,28 +595,25 @@ pub enum BlockHeaderError {
     }
 }
 
+// XXX: used by decode_context_data, which is unused at the moment
```
@tizoc 
please change XXX to Note maybe,
or just remove this comment

bkontur commented on tezos/context/src/lib.rs:
```diff
@@ -1,9 +1,6 @@
 // SPDX-License-Identifier: MIT
 
-// TODO: re-enable this once ocaml-interop has been upgraded
-//#![forbid(unsafe_code)]
```
@tizoc 
we need unsafe here, I think?

tizoc replied:
Yes, good catch.

bkontur commented:
@tizoc 
Bruno, please also check this clippy stuff:

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-1680dd9c5d6642f1ce2af0e56011907f02ebc400d29104c5dbdfe152d738a692R560

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-1680dd9c5d6642f1ce2af0e56011907f02ebc400d29104c5dbdfe152d738a692R777

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-1680dd9c5d6642f1ce2af0e56011907f02ebc400d29104c5dbdfe152d738a692R839

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-a78ca05903fa02ede9a0a128c273f801ac67dc92e7da49929d7eba5a1d73a7a6R612

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-9aaa84afa85b7b6de7ffb7a7d33cc1acd3649d9400ae6bf1a279df85769a3230R22

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-0f623252844b8f4683138376b93fad3990492613f747f64e05d5e0ee259cef9aR617

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-7c77d12ebc8a524986c312be062cf5cf89f29fcadd24ee9fae5cb01fff01c780R82

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-7c77d12ebc8a524986c312be062cf5cf89f29fcadd24ee9fae5cb01fff01c780R92

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-7c77d12ebc8a524986c312be062cf5cf89f29fcadd24ee9fae5cb01fff01c780R111

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-7c77d12ebc8a524986c312be062cf5cf89f29fcadd24ee9fae5cb01fff01c780R121

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-0c75bbf6d1d12c41337143d65e97fd0ad4109b2a721e019c71b576c1faf866fdR38

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-0c75bbf6d1d12c41337143d65e97fd0ad4109b2a721e019c71b576c1faf866fdR90

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-29acf1eb0f8c4990c94336b36cdf08a6a097b7c4b083db42ad6994cdb80cb76bR190

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-c6810b9aa75a706cdd52ec731f9a23f825c3e865ce12f3b69fbbef415b6c2303R7

https://github.com/simplestaking/tezedge/pull/332/files?file-filters%5B%5D=.json&file-filters%5B%5D=.py&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-c787603b11fbe9bcdebbadeca25cd223396ce5e5ffc34e3e32db2cae24a0f247R84

bkontur commented:
@tizoc 
very good, lots of thing looks much simplier now

tizoc commented:
Oops, somehow I missed the notification for the comments on this PR, adressing them now.

tizoc commented:
@bkontur this is rebased against the latest #383, your comments and (most of) the clippy warnings addressed. Also switched the base branch to develop again so that drone runs.

But I see that now the Python tests are failing, not sure why, all used to pass except the last two: http://ci.tezedge.com/simplestaking/tezedge/2332/2/1

And nothing has really changed much on this branch other than keeping it up to date.

How can I get more logs from drone? what is shown on the UI doesn't tell me much.

tizoc commented:
Nevermind my last comment, I found the problem when trying to run the Python tests locally: When I added the new command-line flag to enable/disable computation of the hashes I updated the config file and the reading of that flag, but not the declaration, so this was causing a failure when starting the node. Since the other branch is failing elsewhere for a different reason I never associated this problem with some change there.

Drone is not running now for some reason, but once it does I will know if this was the only issue (I hope).

tizoc commented:
Well, turns out that wasn't it. It is still failing in the same way: http://ci.tezedge.com/simplestaking/tezedge/2671/2/1

tizoc commented:
Well, las change (updating a python constants file I didn't even know existed until now) seems to have solved it.

bkontur commented on light_node/src/configuration.rs:
```diff
@@ -844,17 +842,24 @@ impl Environment {
+                    .expect("Provided value cannot be converted to bool");
+                let compute_context_action_tree_hashes = args
+                    .value_of("compute-context-action-tree-hashes")
+                    .unwrap_or("true")
```
@tizoc 
please change "true" to "false" here

tizoc commented:
Cancelled the last drone build because the last commit is just fixes for clippy warnings, but this one looks like it is about to pass, will merge once it is done: http://ci.tezedge.com/simplestaking/tezedge/2690
